keps/sig-apps/2232-jobs-stopped/README.md

```diff
@@ -0,0 +1,546 @@
+its graceful termination period is honoured. Pods terminated this way are
+considered a failure and the controller does not count terminated Pods towards
+completions. This behaviour is similar to decreasing parallelism to zero.
+
+Similar to existing [JobConditionType](https://github.com/kubernetes/kubernetes/blob/c98f6bf30890f2c5826067ae50cfc36958106e68/staging/src/k8s.io/api/batch/v1/types.go#L167)s
+"Complete" and "Failed", we propose adding a new condition type called
+"Stopped" as a part of the Job's status:
+
+```go
+type JobStatus struct {
+	// The latest available observations of an object's current state. Can be
+	// one of "Stopped", "Failed", or "Complete".
+	Conditions []JobCondition `json:"conditions,omitempty"`
+
+	...
+}
+
+// JobCondition describes current state of a job.
+type JobCondition struct {
+	// Type of job condition, Complete or Failed.
+	Type JobConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+
+	// Last time the condition transit from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	...
+}
+
+type JobConditionType string
+
+// These are valid conditions of a job.
+const (
+	// JobStopped means the job has been stopped.
+	JobStopped JobConditionType = "Stopped"
+	// JobComplete means the job has completed its execution.
+	JobComplete JobConditionType = "Complete"
+	// JobFailed means the job has failed its execution.
+	JobFailed JobConditionType = "Failed"
+)
+```
+
+To determine if a Job has been stopped, users must look for the JobCondition
+with `Type` as "Stopped" and `Status` as true. If such a JobCondition does not
+exist, the Job has never been stopped. Otherwise, if the `Status` is false, the
+Job was previously stopped and then resumed. Inferring this from the Job spec
+is not recommended as the Job controller may not have seen the update yet. When
+a Job is stopped, the Job controller sets the condition only after all Pods are
+terminating or have been deleted.
+
+The Stopped JobCondition also has a `LastTransitionTime` field. This can be
+used to infer how long a Job has been stopped for (if the `Status` value is
+true).
+
+If a Job is created with `stopped: true`, the `StartTime` field of the Job
+be set when it is resumed for the first time.
+
+When a Job is stopped or created in the stopped state, a "Stopped" event is
+recorded. Similarly, when a Job is resumed from its stopped state, a "Resumed"
+event is recorded.
+
+### Test Plan
+
+Unit, integration, and end-to-end tests will be added to test different
+scenarios including:
+
+- Creating a Job with `stopped: true`
+- Stopping an active Job and resuming it again
+- Verifying that partially completed jobs remember completions after a
+  stop-resume cycle
+
+### Graduation Criteria
+
+N/A. Feature will be alpha in the targeted release (1.21).
```

Comment:
Actually, specifying graduation criteria for further steps is desired at the beginning. (we can update that later, but that's the more conscious decision).

Reply:
updated.

Reply:
sorry, I'll let @wojtek-t resolve these